fud2/fud-core/src/run.rs

```diff
@@ -147,13 +162,66 @@ impl EmitBuild for OpEmitData {
             }
         }
 
-        emitter.build_cmd_with_vars(
-            outputs,
-            &self.op_name,
-            inputs,
-            &[],
-            &vars,
-        )?;
+        // Some outputs are created by user specified `shell_deps`. Others are not and have to be
+        // created by running every `shell`/`shell_deps` command.
+        let mut uncreated_outputs: HashSet<String> =
+            HashSet::from_iter((0..outputs.len()).map(|i| {
+                format!("${}", crate::run::io_file_var_name(i, false))
+            }));
```

Comment:
Do you think you could expand on this comment a bit? I understand that some outputs are not created… but why are we tracking them? What is this used for? Is it necessary to craft these special strings that we will then throw away later on? Could we instead keep track of the opposite (the stuff that is created) so we don't have to generate string names for things that we won't need?

Reply:
My mental model with `shell` and `shell_dep` commands is they semantically they are equivalent. Importantly this means that no matter what, every `shell_dep` command is always run, even if it's output is never specified to be used to build the output file. 

The purpose of `shell_dep` commands are to allow Ninja to run code in parallel and better support incremental compilation. 

The purpose of generating these names is so they can be attached to the rules corresponding to `shell_dep` commands which otherwise are never needed. Then the file to be created can depend on these rules.

On "why are some of these string being thrown away." There is no reason, it looks possible to just directly generate the correct set of names all at once. I should just do that.